src/sftp.rs

```diff
@@ -0,0 +1,400 @@
+            dir
+        } else {
+            // they're trying to write to / itself?
+            return Err(Error::Remote(io::Error::new(
+                io::ErrorKind::PermissionDenied,
+                "trying to write to root as a file",
+            )));
+        };
+
+        let test = self
+            .session
+            .command("test")
+            .arg("(")
+            .arg("-d")
+            .arg(dir)
+            .arg(")")
+            .arg("-a")
+            .arg("(")
+            .arg("-w")
+            .arg(dir)
+            .arg(")")
+            .status()?;
+
+        if test.success() {
+            Ok(())
+        } else if self
+            .session
+            .command("test")
+            .arg("-e")
+            .arg(path)
+            .status()?
+            .success()
+        {
+            Err(Error::Remote(io::Error::new(
+                io::ErrorKind::PermissionDenied,
+                "parent directory not writeable",
+            )))
+        } else {
+            Err(Error::Remote(io::Error::new(
+                io::ErrorKind::NotFound,
+                "parent directory does not exist",
+            )))
+        }
+    }
+
+    /// Initialize the given operation on the target file.
+    ///
+    /// Note that this function is not guaranteed to be side-effect free for writes. Specifically,
+    /// it will create the target file in order to test whether it is writeable.
+    fn init_op(&mut self, mode: Mode, path: impl AsRef<std::path::Path>) -> Result<(), Error> {
+        if mode.is_write() {
+            // for writes, we want a stronger (and cheaper) test than can()
+            // we can't actually use `touch`, since it also works for dirs
+            // note that this works b/c stdin is Stdio::null()
+            let touch = self
+                .session
+                .command("cat")
+                .arg(">>")
+                .arg(path.as_ref())
+                .stdin(Stdio::null())
+                .output()?;
+            if touch.status.success() {
+                return Ok(());
+            }
+
+            // let's find out _why_ it failed
+            let stderr = String::from_utf8_lossy(&touch.stderr);
+            if stderr.contains("No such file or directory") {
+                Err(Error::Remote(io::Error::new(
+                    io::ErrorKind::NotFound,
+                    &*stderr,
+                )))
+            } else {
+                Err(Error::Remote(io::Error::new(
+                    io::ErrorKind::PermissionDenied,
+                    &*stderr,
+                )))
+            }
+        } else {
+            self.can(Mode::Read, path)
+        }
+    }
+
+    /// Open the remote file at `path` for writing.
+    ///
+    /// If the remote file exists, it will be truncated. If it does not, it will be created.
+    ///
+    /// Note that errors may not propagate until you call [`close`], including if the remote file
```

Comment:
Might be worth mentioning the `can()` check here.
Ideally, this whole module would benefit from a small doc example I think.

Reply:
So, `can` is actually _less_ reliable than the checks that these methods now do internally, at least for writes. If you _also_ did a `can` call, you wouldn't get any more errors.

Reply:
Yeah, I think it would be good to try to give some guidance on what `can()` does vs what is provided by `write_to` internals. (i.e., when should I call each one)

Reply:
How's ec726ba43c4d2881109b610aa7cf2a8be212e741 ?